cinema/views.py

```diff
@@ -56,3 +83,66 @@ def get_serializer_class(self):
             return MovieSessionDetailSerializer
 
         return MovieSessionSerializer
+
+    def get_queryset(self):
+        queryset = self.queryset
+
+        if self.action == "retrieve":
+            queryset = queryset.select_related("movie")
+
+        elif self.action == "list":
+            queryset = (
+                queryset.select_related("cinema_hall").annotate(
+                    tickets_available=(
+                        F("cinema_hall__rows") * F("cinema_hall__seats_in_row")
+                        - Count("tickets")
+                    )
+                )
+            ).order_by("id")
+
+        movie = self.request.query_params.get("movie")
+        date = self.request.query_params.get("date")
+        if date:
+            queryset = queryset.filter(show_time__date=date)
+        if movie:
+            queryset = queryset.filter(movie__id=movie)
+        if self.action in ("list", "retrieve"):
+            queryset = queryset.select_related("movie", "cinema_hall")
```

Comment:
The `select_related` call for 'movie' and 'cinema_hall' is already handled in the 'retrieve' action. You can remove this redundant call within the 'list' and 'retrieve' check to avoid unnecessary repetition.